Add support for non-bzImage kernels

Jannik2099 opened the pull request:
Many non-x86 platforms do not use / support the bzImage format, but use raw binaries instead.
Sadly there is no metadata header that contains the version string, nor is there any standardized address within the image (that I'm aware of).
Furthermore, on these platforms the kernel image doesn't decompress itself but is compressed as a whole, with the task of decompression being left to the bootloader.

This pull adds support for uncompressed and compressed (currently gzip, bzip2, xz, lzma, lz4, lzo and zstd) raw kernel images.
This is implemented by parsing the raw file for the string "Linux version $VERSION ...", which is the greeter line you see on boot.

For this the following new dependencies are needed:
sys-apps/file[python]
dev-python/lz4
dev-python/python-lzo
dev-python/zstandard

Jannik2099 commented:
Just noticed the CI fails because of two reasons:
the magic module from file[python] doesn't seem to exist, and it's not packaged in pip - is there any way to fix this or should I replace the dependency with another libmagic implementation?
the lzo module fails to build due to missing headers - again I'm not familiar with semaphore, any advice welcome

mgorny commented:
Technically speaking, reading version number was supposed to be externalized to [kernel-readver](https://github.com/mgorny/kernel-readver) but this never has been finished. Since there is apparently no trivial way to grab it (except for grepping like you do), I'm not sure if that still makes sense.

In any case, please don't use libmagic. All compressors should be able to clearly determine (e.g. by raising an exception) whether they get a correct file or not. Also, please make sure to throw a user-visible error only if the missing compression format package is actually needed.

Jannik2099 commented:
> Also, please make sure to throw a user-visible error only if the missing compression format package is actually needed.

not sure if I understand you - you want me to throw an error if the decompressor is not installed, but to find out the compression format I'd either need libmagic, or said decompressor.
I could throw a generic "Format not recognized" if all available decompressors fail, but I don't see how I'd be able to find out the exact format without libmagic?

mgorny commented:
I dare say all of them have some trivial magic bytes you can detect.

Jannik2099 commented:
They all do indeed have some trivial magic bytes - finding them was not so trivial though.
Fixed, and also fixed CI. An exception is thrown if the compression format is recognized but the required decompressor not installed - is this to your liking?

mgorny commented on ecleankernel/file.py:
```diff
@@ -91,6 +95,60 @@ def __init__(self,
 
+    def __decompress_raw(self) -> bytes:
+        f = open(self.path, 'rb')
+        magic = [
```
Please use a dict and don't rely on hardcoding indices. E.g.:

```
magic = {
    b'...': 'gzip',
    ...
```

Then use a `for` loop to determine which one matched, and go forward from these strings.

Or you could go a bit further and use strings matching module names, and then for most of the options you'd use something like:

```
mod = importlib.import_module(name)
b = getattr(mod, 'decompress')(f.read())
```

and you'd probably only need an exception for zstd.

Jannik2099 replied:
That's exactly what I wanted to do, but I didn't know you could dynamically import and call functions by variable - I still have a lot of python to learn, thanks!

mgorny commented on ecleankernel/file.py:
```diff
@@ -101,18 +159,37 @@ def read_internal_version(self) -> str:
+            # unlike with bzImage, the raw kernel binary has no header
+            # that includes the version, so we parse the version message
+            # that appears on boot
+            s = b.decode('utf8', 'ignore')
```
Maybe ASCII would be a simpler choice here.

Jannik2099 replied:
It is indeed, was about 20% faster on my system

mgorny commented on test/test_file.py:
```diff
@@ -24,20 +25,55 @@ def write_bzImage(path: Path,
+    with gzip.open(path, 'wb') as f:
+        # gzip would compress a string of 0s below 0x200,
+        # so we fill in some incompressible gibberish
+        f.write(os.urandom(0xffff))
```
This doesn't really guarantee it will be 'incompressible'. It would be better to use some predictable algorithm, like playing with `range()`.

mgorny commented on ecleankernel/file.py:
```diff
@@ -7,6 +7,7 @@
 import os
 import shutil
 import struct
+import importlib
```
Please sort includes.

mgorny commented on ecleankernel/file.py:
```diff
@@ -91,6 +96,42 @@ def __init__(self,
         super().__init__(path, KernelFileType.KERNEL)
         self.internal_version = self.read_internal_version()
 
+    def __decompress_raw(self) -> bytes:
```
Double underscore is for special Python methods, just name it normally here.

Jannik2099 replied:
I thought __ would mark "private" functions - fixed

mgorny commented on ecleankernel/file.py:
```diff
@@ -91,6 +96,42 @@ def __init__(self,
+                    import zstandard
+                    b = zstandard.ZstdDecompressor().decompress(f.read())
+                else:
+                    b = getattr(mod, 'decompress')(f.read())
```
```suggestion
                    return getattr(mod, 'decompress')(f.read())
```

You don't really have to iterate over all of them, esp. that magics don't overlap.

mgorny replied:
For extra kudos, you could try seeing if all of them can have some streaming classes (i.e. `open()` function), so we could avoid loading the whole file into memory unnecessarily.

Jannik2099 replied:
I somehow didn't think of just returning them - oops

Sadly, lzo only provides `compress()` and `decompress()` functions :/

mgorny replied:
Ok, I suppose this makes converting others pointless.

mgorny commented on ecleankernel/file.py:
```diff
@@ -101,18 +142,38 @@ def read_internal_version(self) -> str:
+            # unlike with bzImage, the raw kernel binary has no header
+            # that includes the version, so we parse the version message
+            # that appears on boot
+            s = b.decode('ascii', 'ignore')
```
Actually, you don't have to do that at all. Just search for `b'Linux version '` and work on bytes until the very end. Looking at it, the existing code decodes too early indeed.

mgorny commented on ecleankernel/file.py:
```diff
@@ -101,18 +142,38 @@ def read_internal_version(self) -> str:
+                    f'Kernel file {self.path} does not appear '
+                    f'to have a version string, '
+                    f'or the compression format was not recognized')
+            pos = pos + len(ver_start)
```
```suggestion
            pos += len(ver_start)
```

mgorny commented on ecleankernel/file.py:
```diff
@@ -101,18 +142,38 @@ def read_internal_version(self) -> str:
+                    f'to have a version string, '
+                    f'or the compression format was not recognized')
+            pos = pos + len(ver_start)
+            sbuf = s[pos:pos + 0x100]
```
Is there a reason for this `0x100`? In the above code, the idea was not to have to read the whole file into memory but you do that already anyway.

Jannik2099 replied:
Now that we only decode as needed, there is indeed a reason for it :)

mgorny commented on ecleankernel/file.py:
```diff
@@ -91,6 +96,38 @@ def __init__(self,
+            b'\x5d\x00\x00': 'lzma',
+            b'\x04\x22\x4d\x18': 'lz4.frame',
+            b'\x28\xb5\x2f\xfd': 'zstandard',
+            b'\x89\x4c\x5a\x4f\x00\x0d\x0a\x1a\x0a': 'lzo'
```
Nit because I'll probably miss it later. It's generally a good idea to add extra `,` to make smaller diffs when adding stuff.

```suggestion
            b'\x89\x4c\x5a\x4f\x00\x0d\x0a\x1a\x0a': 'lzo',
```

mgorny commented on ecleankernel/file.py:
```diff
@@ -101,18 +138,37 @@ def read_internal_version(self) -> str:
+                raise UnrecognizedKernelError(
+                    f'Kernel file {self.path} terminates before expected '
+                    f'version string position ({offset + 0x200})')
+            ret = buf.decode(errors='ignore').split(' ', 1)
```
This and below to:

```suggestion
            ret = buf.split(b' ', 1)
```

mgorny commented on ecleankernel/file.py:
```diff
@@ -101,18 +138,37 @@ def read_internal_version(self) -> str:
-        return buf.split(b' ', 1)[0].decode()
+                f'Kernel file {self.path} terminates '
+                f'before end of version string')
+        return (ret[0])
```
...and then:

```suggestion
        return ret[0].decode('ASCII')
```

I suppose we shouldn't ignore errors at this point as kernel version should be plain ASCII, right?

Jannik2099 replied:
Very good change - indeed no need to do errors=ignore - I omitted specifying the encoding now, couldn't measure a speed difference for the just 0x100 bytes we decode now

mgorny commented on test/test_file.py:
```diff
@@ -24,20 +28,60 @@ def write_bzImage(path: Path,
             KernelImage(path).read_internal_version(),
             '1.2.3')
 
+    def test_read_internal_version_raw(self) -> None:
```
Would be nice to move this into proper pytest-based case and parametrize it for different compressors.

Jannik2099 replied:
Sadly my knowledge of pytest only covers cursing at its output, no idea what you'd like to see here

mgorny commented:
If I'm asking for too much, please let me know. I suppose I could also do some groundwork before your patch, e.g. rewriting tests to use pytest.

Jannik2099 commented:
> If I'm asking for too much, please let me know.
Quite the opposite, I'm learning a lot here and your feedback is very helpful!

mgorny commented:
Ok, I think this is good enough now. Is it ready to merge or are you still working on something?

Jannik2099 commented:
No, I'm done here. Can you give it a quick test on x86 before merging? I only have access to arm64 hardware right now.

mgorny commented:
I'm in the same situation.

mgorny commented:
Since this is a major change, I'd use your signed-off-by for copyright.

Jannik2099 commented:
> Since this is a major change, I'd use your signed-off-by for copyright.

Sure! I'm a tad of a git noob though, does this require me to squash this into one commit and sign it off myself?

mgorny commented:
Yes, please. `git rebase -i origin/master` usually works.

mgorny commented:
and `--signoff`.

Jannik2099 commented:
wait, was this correct or do I need to force push to remove the old commits?

mgorny commented:
Yes, you need to force push

Jannik2099 commented:
Heh, learned both python AND git in a week - thanks!

mgorny commented:
Your signoff should contain your real name.

mgorny reviewed (approved):
Thank you!
